src/url.ts

```diff
@@ -0,0 +1,56 @@
+/**
+ * URL related helper functions
+ */
+
+/**
+ * Converts various input formats to URL objects.
+ * Handles URL objects, domain strings, relative paths, and full URLs.
+ * @param {string|URL} input - Input to convert to URL object
+ * @returns {URL|null} URL object or null if input is invalid
+ */
+export const toURL = (input: string | URL | null | undefined): URL | null => {
+  if (!input) return null;
+  if (input instanceof URL) return input;
+
+  if (
+    typeof input === 'string' &&
+    !input.includes('://') &&
+    !input.startsWith('/')
+  ) {
+    return new URL(`https://${input}`);
+  }
+
+  if (typeof input === 'string' && input.startsWith('/')) {
+    return new URL(input, window.location.origin);
+  }
+
+  try {
+    return new URL(input as string);
+  } catch (error) {
+    return null;
+  }
```

Comment:
We don't need this `try { } catch`, or else, let's add the whole block in `try { } catch` block